lib/common/udp_tunnel.c

```diff
@@ -0,0 +1,208 @@
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <unistd.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+
+#include "h2o/httpclient.h"
+
+struct st_h2o_udp_tunnel_t {
+    h2o_tunnel_t super;
+    socklen_t len;
+    h2o_socket_t *sock;
+    h2o_loop_t *loop;
+    h2o_buffer_t *inbuf; /* for datagram fragments */
+};
+
+static void tunnel_on_destroy(h2o_tunnel_t *_tunnel)
+{
+    struct st_h2o_udp_tunnel_t *tunnel = (void *)_tunnel;
+
+    h2o_buffer_dispose(&tunnel->inbuf);
+    h2o_socket_close(tunnel->sock);
+    free(tunnel);
+}
+
+static void tunnel_socket_on_read(h2o_socket_t *sock, const char *err)
+{
+    struct st_h2o_udp_tunnel_t *tunnel = sock->data;
+
+    if (err != NULL) {
+        tunnel->super.on_read(&tunnel->super, err, NULL, 0);
+        return;
+    }
+
+    uint8_t buf[16384];
+    struct msghdr mess;
+    struct sockaddr sa;
+    struct iovec iov;
+    memset(&mess, 0, sizeof(mess));
+    mess.msg_name = &sa;
+    mess.msg_namelen = sizeof(sa);
+    iov.iov_base = buf;
+    iov.iov_len = sizeof(buf);
+    mess.msg_iov = &iov;
+    mess.msg_iovlen = 1;
+    ssize_t rret;
+    while ((rret = recvmsg(h2o_socket_get_fd(tunnel->sock), &mess, 0)) == -1 && errno == EINTR)
+        ;
+
+    h2o_socket_read_stop(tunnel->sock);
+
+    if (tunnel->super.on_udp_read) {
+        h2o_iovec_t hiov = h2o_iovec_init(iov.iov_base, iov.iov_len);
+        tunnel->super.on_udp_read(&tunnel->super, NULL, &hiov, 1);
+    } else {
+        struct {
+            uint8_t buf[8];
+            uint8_t *end;
+        } varint;
+        h2o_iovec_t vec[3];
+        const uint8_t zero = 0;
+
+        varint.end = ptls_encode_quicint(varint.buf, iov.iov_len);
+        vec[0].base = (char *)&zero;
+        vec[0].len = sizeof(zero);
+        vec[1].base = (char *)varint.buf;
+        vec[1].len = varint.end - varint.buf;
+        vec[2] = h2o_iovec_init(iov.iov_base, iov.iov_len);
+        tunnel->super.on_read(&tunnel->super, NULL, vec, sizeof(vec) / sizeof(vec[0]));
+    }
+
+    return;
+}
+
+h2o_iovec_t get_next_chunk(const uint8_t *bytes, size_t len, size_t *to_consume, int *skip)
```

Comment:
I think we can reduced the information being returned to the caller from three elements to two elements, by changing the signature of the function to: `size_t get_next_chunk(const uint8_t *bytes, size_t len, h2o_iovec_t *chunk)`.

The only case where the vector has to be examined is when some bytes are to be consumed. And a NULL chunk with some bytes being consumed can indicate the fact that a chunk has been skipped.